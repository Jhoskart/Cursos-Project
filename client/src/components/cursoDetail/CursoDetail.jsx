import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { deleteCursoRequest, updateCursoRequest } from '../../redux/slice/thunks'
import { Button, Modal, ModalHeader, ModalBody, ModalFooter, FormGroup, Input, Form } from 'reactstrap'
import './cursoDetail.css'

export default function CursoDetail({ name, description, id, image }) {
    const [modal, setModal] = useState(false)
    const dispatch = useDispatch()
    
    const switchModal = () => {
        setModal(!modal)
    }

    const deleteCurso = () => {
        dispatch(deleteCursoRequest(id))
        setModal(false)
    }

    ////////// Editar Curso //////////

    const [editCurso, setEditCurso] = useState(null)

    const [editCursoData, setEditCursoData] = useState(
        name && description ? { name, description, image } : {
            name: '',
            description: '',
            image: ''
        }
    )

    const handleEditCurso = (e, cursoId) => {
        e.preventDefault()
        setEditCurso(cursoId)
    }

    const handleFormChange = (e) => {
        setEditCursoData({
            ...editCursoData,
            [e.target.name]: e.target.value
        })
    }

    const handleFormSubmit = (e) => {
        e.preventDefault()
        dispatch(updateCursoRequest(editCurso, editCursoData))
        setEditCurso(null)
        setEditCursoData({
            name: '',
            description: '',
            image: ''
        })
        setModal(false)
        
    }

    const modalStyles = {
        position: 'relative',
        top: '50%',
        left: '15%',
        transform: 'translate(-50%, -50%)'
    }

    const cancelEdit = () => {
        setEditCurso(null)
        setModal(!modal)
    }


    return (
        <div>
            <Button onClick={switchModal} color='light' >Mas informacion aqui</Button>
                <Modal style={modalStyles} isOpen={modal} >
                    <Form onSubmit={handleFormSubmit}>
                        {editCurso === id ? (
                            <ModalHeader>
                                <FormGroup>
                                    <Input 
                                    type='text' 
                                    name='name' 
                                    id='name' 
                                    placeholder='Nombre del curso' 
                                    value={editCursoData.name}
                                    onChange={handleFormChange}
                                    />
                                </FormGroup>
                            </ModalHeader>
                        ) : (
                            <ModalHeader>{name}</ModalHeader>
                        )}
                    
                        {editCurso === id ? (
                            <ModalBody>

                                {editCursoData.image ? (
                                    <img className='boxForImage' src={editCursoData.image} alt='Imagen del curso' />
                                ) : (
                                    <div className='boxForImage'>
                                        <p className='margin' >Carga tu imagen</p>
                                    </div>
                                )
                                    }

                                <FormGroup>
                                    <Input
                                    type='url'
                                    name='image'
                                    id='image'
                                    placeholder='Url de la imagen'
                                    value={editCursoData.image}
                                    onChange={handleFormChange}
                                    />
                                </FormGroup>

                                <FormGroup>
                                    <Input
                                    type='textarea'
                                    name='description'
                                    id='description'
                                    placeholder='Descripcion del curso'
                                    value={editCursoData.description}
                                    onChange={handleFormChange}
                                    rows='14'   
                                    />
                                </FormGroup>

                            </ModalBody>
                            ) : (
                            <ModalBody>
                                <img className='boxForImage2' src={image} alt="imagen del curso" />
                                <div>
                                    <p>{description}</p>
                                </div>
                            </ModalBody>
                            )}
                            {editCurso === id ? (
                            <ModalFooter>
                                <Button color='primary' type='submit' >
                                Guardar
                                </Button>
                                <Button color='secondary' onClick={cancelEdit}>
                                    Cancelar
                                </Button>
                            </ModalFooter>
                            ) : (
                                <ModalFooter>
                                    <Button color='danger' onClick={() => deleteCurso(id)} >Eliminar</Button>
                                    <Button color='primary' onClick={(e) => handleEditCurso(e, id)} >Editar</Button>
                                    <Button onClick={switchModal} color='secondary'>Cancelar</Button> 
                                </ModalFooter> 
                            )}
                    </Form>
                </Modal>
        </div>
    )
}
