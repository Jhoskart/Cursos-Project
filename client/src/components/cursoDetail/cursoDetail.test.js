import {screen, render, fireEvent} from '@testing-library/react';
import { Provider } from 'react-redux';
import * as ReactRedux from 'react-redux';
import CursoDetail from './CursoDetail'
import { configure, shallow, mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";
import configureStore from 'redux-mock-store';
import axios from 'axios';
import nock from 'nock';
import isReact from 'is-react';
import thunk from 'redux-thunk';
import * as data from '../../db.json'
import { MemoryRouter } from 'react-router-dom'
axios.defaults.adapter = require('axios/lib/adapters/http');


configure({ adapter: new Adapter() });

describe('<CursoDetail />', () => {
    let cursoDetail, state, store;
    const mockStore = configureStore([thunk])
    let cursos = data.cursos
    state = { 
        cursos: []
    }
    store = mockStore(state)

    beforeEach(() => {

        const apiMock = nock('http://localhost:3001').persist();

        apiMock.get("/cursos").reply(200, data.cursos);

        cursoDetail = (cursos) => 
        mount(
            <Provider store={store}>
                    <CursoDetail 
                        name={cursos.name}
                        description={cursos.description}
                        image={cursos.image}
                        id={cursos.id}
                    />
            </Provider>
        )
    });

    afterEach(() => {
        jest.restoreAllMocks();
    }),

    it('should render correctly', () => {
        const { container } = render(
            <Provider store={store}>
                <CursoDetail 
                    name={cursos.name}
                    description={cursos.description}
                    image={cursos.image}
                    id={cursos.id}
                />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    })

    describe('Estructura', () => {

        it('shoulds render a div contenedor', () => {
            const { container } = render(
                <Provider store={store}>
                    <CursoDetail
                        name={cursos.name}
                        description={cursos.description}
                        image={cursos.image}
                        id={cursos.id}
                    />
                </Provider>
            );
            expect(container.querySelector('div')).toBeInTheDocument();
        })

        it('shouls render a curso image', () =>{
            const { container } = render(
                <Provider store={store}>
                    <CursoDetail
                        name={cursos.name}
                        description={cursos.description}
                        image={cursos.image}
                        id={cursos.id}
                    />
                </Provider>
            );
            fireEvent.click(screen.getByText('Mas informacion aqui'));
            expect(screen.getByAltText('imagen del curso')).toBeInTheDocument(); 
        })

        it('shoulds render a form to edit', () =>{
            const { container } = render(
                <Provider store={store}>
                    <CursoDetail
                        name={cursos.name}
                        description={cursos.description}
                        image={cursos.image}
                        id={cursos.id}
                    />
                </Provider>
            );
            fireEvent.click(screen.getByText('Mas informacion aqui'));
            fireEvent.click(screen.getByText('Editar'));
            expect(screen.getByPlaceholderText('Nombre del curso')).toBeInTheDocument();
            expect(screen.getByPlaceholderText('Descripcion del curso')).toBeInTheDocument();
            expect(screen.getByPlaceholderText('Url de la imagen')).toBeInTheDocument();
        })

        it('should render the curso name in h5', () =>{
            const { container } = render(
                <Provider store={store}>
                    <CursoDetail
                        name={cursos.name}
                        description={cursos.description}
                        image={cursos.image}
                        id={cursos.id}
                    />
                </Provider>
            );
            fireEvent.click(screen.getByText('Mas informacion aqui'));
            expect(screen.getByRole('heading')).toBeInTheDocument();
        })
        
    })
})

