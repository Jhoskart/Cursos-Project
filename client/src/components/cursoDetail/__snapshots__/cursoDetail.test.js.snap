// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CursoDetail /> should render correctly 1`] = `
<div>
  <div>
    <button
      class="btn btn-light"
      type="button"
    >
      Mas informacion aqui
    </button>
  </div>
</div>
`;
